### PR TITLE
Adjust QA scripts remove installing pip whl from direct links as pip will disregard the "-f" option in that case

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
Reduce QA tests are redownloading pytorch, because it was specified from a url

#### What happened in this PR?
Modify QA test utilities to first try to download pytorch from downloaded pip packages first and only if that doesn't work try to fetch it from the url

**JIRA TASK**: [DALI-XXXX]

### CODE REVIEW

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

File: qa/setup_packages.py [hunk 1]
```diff
@@ -62,67 +69,79 @@ def get_pyvers_name(name, cuda):
...
-            ret.append(get_pyvers_name(val, cuda))
-        else:
-            ret.append(key + "==" + val)
-        variant = variant // len(get_package(packages, key, cuda))
+        pkg_list_len = len(get_package_list(package_data, key, cuda))
+        idx = variant % pkg_list_len
+        val = get_package_list(package_data, key, cuda)[idx]
+        pkg_str = key
+        if isinstance(val, str):
+            if val.startswith('http'):
```
Comment (JanuszL) [inline qa/setup_packages.py]: Do we need that case anymore?

File: qa/setup_packages.py [hunk 2]
```diff
@@ -62,67 +69,79 @@ def get_pyvers_name(name, cuda):
              pass
     return ""
 
+def get_version(package_version_entry):
+    if isinstance(package_version_entry, tuple):
```
Comment (JanuszL) [inline qa/setup_packages.py]: Just an idea. Maybe instead of tuple we should create some simple class to keep that data.

File: qa/test_template.sh [hunk 3]
```diff
@@ -39,34 +38,36 @@ epilog=${epilog-:}
...
         test_body
 
         # remove packages
-        remove=$($topdir/qa/setup_packages.py -r  -u $pip_packages --cuda ${CUDA_VERSION})
-        if [ -n "$remove" ]
-        then
-        pip uninstall -y $remove
-        fi
+        eval "remove=(`$topdir/qa/setup_packages.py -r  -u $pip_packages --cuda ${CUDA_VERSION} | tr -d '[],'`)"
+        for pkg in "${packages[@]}"; do
```
Comment (JanuszL) [inline qa/test_template.sh]: Why loop?

File: qa/download_pip_packages.sh [hunk 4]
```diff
@@ -10,13 +10,11 @@ do
...
-    do
-        inst=$(python setup_packages.py -i $i -u $pip_packages --cuda ${CUDA_VERSION})
-        if [ -n "$inst" ]
-        then
-            pip download $inst -d /pip-packages
-        fi
+    last_config_index=`./setup_packages.py -n -u $pip_packages --cuda ${CUDA_VERSION}`
+    for i in `seq 0 $last_config_index`; do
+        eval "packages=(`./setup_packages.py -i $i -u $pip_packages --cuda ${CUDA_VERSION} --include-link | tr -d '[],'`)"
+        for pkg in "${packages[@]}"; do
```
Comment (JanuszL) [inline qa/download_pip_packages.sh]: Why loop?

File: qa/test_template.sh [hunk 5]
```diff
@@ -39,34 +38,36 @@ epilog=${epilog-:}
...
+            pkg="${packages[${j}]}"
+            pkg_with_link="${packages_with_link[${j}]}"
 
-        # If we just installed tensorflow, we need to reinstall DALI TF plugin
-        if [[ "$inst" == *tensorflow-gpu* ]]; then
-            pip uninstall -y nvidia-dali-tf-plugin || true
-            pip install /opt/dali/nvidia-dali-tf-plugin*.tar.gz
-        fi
-        fi
+            pip install "$pkg" -f /pip-packages --no-index || pip install "$pkg_with_link"
```
Comment (JanuszL) [inline qa/test_template.sh]: So `"$pkg_with_link"` doesn't work with `-f /pip-packages`, right?

**Comment by JanuszL** [comment]
> Adjust QA scripts to avoid redownloading torch and torchvision when unnecessary

Maybe you should reword it to: `Adjust QA scripts remove installing pip whl from direct links as pip will disregard the "-f" option in that case`, or something like that. The issue is not torch specific, even if it currently manifests in this particular case.
Same to the PR description.

File: qa/download_pip_packages.sh (hunk 4 above)
Comment (jantonguirao) [reply qa/download_pip_packages.sh]: now the output of this script is an array, as opposed to a string

File: qa/test_template.sh (hunk 3 above)
Comment (jantonguirao) [reply qa/test_template.sh]: now the output of this script is an array, as opposed to a string

File: qa/setup_packages.py (hunk 2 above)
Comment (jantonguirao) [reply qa/setup_packages.py]: I think it's an overkill (for now). If I third argument goes into place I might consider it

File: qa/setup_packages.py (hunk 1 above)
Comment (jantonguirao) [reply qa/setup_packages.py]: We don't but I was hesitant to delete it just in case it turns out that we need it any time soon. Removing it wouldn't affect anything though

File: qa/test_template.sh (hunk 5 above)
Comment (jantonguirao) [reply qa/test_template.sh]: Here we are doing :
1. try to install from /pip-packages and only that (no index), if that fails then pull from the index or the additional url

**Comment by jantonguirao** [comment]
> > Adjust QA scripts to avoid redownloading torch and torchvision when unnecessary
> 
> Maybe you should reword it to: `Adjust QA scripts remove installing pip whl from direct links as pip will disregard the "-f" option in that case`, or something like that. The issue is not torch specific, even if it currently manifests in this particular case.
> Same to the PR description.

Changed it

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build
